collections/_article/d3d12-shaders-with-a-twist.md

```diff
@@ -0,0 +1,112 @@
+---
+title: "Direct3D 12: Shaders with a twist"
+excerpt: "Godot is about to get a Direct3D 12 rendering driver. This is a discussion on the exotic approaches taken to face some of the challenges posed in the area of shaders."
+categories: ["progress-report"]
+author: Pedro J. Estébanez
+image: /storage/blog/covers/d3d12-shaders-with-a-twist.jpg.jpg
+date: 2023-04-12 20:00:00
+---
+
+This is Pedro, a.k.a. [RandomShaper](https://twitter.com/RandomPedroJ) in the Godot community.
+
+You may already know what I have been implementing a **Direct3D 12 rendering driver** for Godot. Since [the announcement](https://w4games.com/2022/08/12/w4-games-donates-direct3d-12-support-to-godot/) that W4 Games donated my time working on that to the project, the code has received an amount of maintenance. In fact, [the pull request](https://github.com/godotengine/godot/pull/70315) hasn't still been merged since there has been a continuous flow of improvements.
+
+In this article I'd like to discuss specifically how I faced some of the challenges posed by different aspects of making shaders work, which was by far the most technically interesting area. Be warned that this piece is very dense in technical information.
+
+## Translation of GLSL/SPIR-V to DXIL
+
+Godot rendering architecture is built with the industry standard Vulkan graphics API in mind. And that is great. However, once you want to plug new rendering APIs into the engine, you'll find that some elements need to be wired in a manner that they work with the alternative API (Direct3D 12 in this case.)
+
+Shaders in Godot are written in its custom, very **GLSL**-like language. (The builtin shaders are already GLSL in the source code.) Shader code in this language is compiled by the engine (thanks to **glslang**) into SPIR-V. which is a well-defined format that stores the binary representation of a shader.
+
+SPIR-V and Vulkan play nice. However, Direct3D 12 features its own binary shader format: **DXIL**. This format is structured as a container with its most important contained chunks being shader metadata and the compiled shader code in the form of **LLVM bitcode** (yes, the same LLVM/Clang uses for an intermediate representation of the compiled code.) As a consequence, this rendering driver would need a way to generate DXIL from either the GLSL or the SPIR-V representations of the Godot shaders.
+
+### First round: SPIRV-Cross
+
+After assessing the available third-party libraries and open source projects capable of doing that (and license-compatible), I found **SPIRV-Cross** was the most sensible choice: it would translate GLSL to HLSL (the high level counterpart of GLSL in the DirectX world). From that point, the D3D12 renderer would use the standard **DirectX Shader Compiler** library to compile HLSL into DXIL.
```

Comment:
```suggestion
After assessing the available third-party libraries and open source projects capable of doing that (and license-compatible), I found **SPIRV-Cross** was the most sensible choice: it would translate GLSL to HLSL (the high-level counterpart of GLSL in the DirectX world). From that point, the D3D12 renderer would use the standard **DirectX Shader Compiler** library to compile HLSL into DXIL.
```